crates/ide-assists/src/handlers/wrap_return_type.rs

```diff
@@ -26,6 +42,61 @@ use crate::{AssistContext, AssistId, AssistKind, Assists};
 // fn foo() -> Result<i32, ${0:_}> { Ok(42i32) }
 // ```
 pub(crate) fn wrap_return_type_in_result(acc: &mut Assists, ctx: &AssistContext<'_>) -> Option<()> {
+    wrap_return_type(acc, ctx, WrapperKind::Result)
+}
+
+enum WrapperKind {
+    Option,
+    Result,
+}
+
+impl WrapperKind {
+    fn assist_id(&self) -> AssistId {
+        let s = match self {
+            WrapperKind::Option => "wrap_return_type_in_option",
+            WrapperKind::Result => "wrap_return_type_in_result",
+        };
+
+        AssistId(s, AssistKind::RefactorRewrite)
+    }
+
+    fn label(&self) -> &'static str {
+        match self {
+            WrapperKind::Option => "Wrap return type in Option",
+            WrapperKind::Result => "Wrap return type in Result",
+        }
+    }
+
+    fn happy_ident(&self) -> &'static str {
+        match self {
+            WrapperKind::Option => "Some",
+            WrapperKind::Result => "Ok",
+        }
+    }
+
+    fn core_type(&self, famous_defs: FamousDefs<'_, '_>) -> Option<hir::Enum> {
+        match self {
+            WrapperKind::Option => famous_defs.core_option_Option(),
+            WrapperKind::Result => famous_defs.core_result_Result(),
+        }
+    }
+
+    fn symbol(&self) -> hir::Symbol {
+        match self {
+            WrapperKind::Option => hir::sym::Option.clone(),
+            WrapperKind::Result => hir::sym::Result.clone(),
+        }
+    }
+
+    fn wrap_type(&self, type_ref: &ast::Type) -> ast::Type {
+        match self {
+            WrapperKind::Option => make::ext::ty_option(type_ref.clone()),
+            WrapperKind::Result => make::ext::ty_result(type_ref.clone(), make::ty_placeholder()),
+        }
+    }
+}
+
+fn wrap_return_type(acc: &mut Assists, ctx: &AssistContext<'_>, kind: WrapperKind) -> Option<()> {
```

Comment:
The way this is setup we do this work twice now (always), might be better to only have one entry point, do the filtering logic once and then add both assists at once. We also probably want to group the two assist under one assist group (label being something "Wrap return type in ...").